README
This is a lynx dump of http://www.tana.it/sw/zdkimfilter/

   home [1]old
   documentation:[2]zdkimfilter [3]dkimsign [4]redact [5]zaggregate
   [6]zfilter_db [7]zdkimfilter.conf

                                  ZDKIMFILTER
      "z" DKIM filter for Courier-MTA using the OpenDKIM Project's library

Download


   [8][rel.notes]
   [9]current--> v1.4 Thu 08 Jan 2015 16:01:37 [10][gpg signature]
   [11][announce] [12][rel.notes]
   [13]v1.3 Tue 23 Dec 2014 17:12:41 [14][gpg signature] [15][announce]
   [16][rel.notes]
   [17]v1.2 Sat 30 Mar 2013 15:03:19 [18][gpg signature] [19][announce]
   [20][rel.notes]
   [21]v1.1 Wed 05 Dec 2012 17:12:44 [22][gpg signature] [23][announce]
   [24][rel.notes]
   [25]v1.0 Sun 28 Oct 2012 20:10:41 [26][gpg signature] [27][announce]

Requirements

     * courier and libopendkim, of course,
     * libopendbx,
     * libnettle,
     * libresolv,
     * libidn2*,
     * libunistring*,
     * zlib*,
     * uuid*,
     * the pkg-config utility, and
     * libtool (even if we build no libraries.)
     * In addition, the Public Suffix List* is used if configured.

   (*) New in version 1.5

   (Some [28]old mumbling about where to get required software may still
   be valid.)

License

   [29]GPLv3

   Keep in mind that DKIM relies on patented technology that may require a
   separate licence.

   As far as software copyright is concerned, zdkimfilter is free
   software: you can redistribute it and/or modify it under the terms of
   the GNU [30]General Public Licence as published by the Free Software
   Foundation, either version 3 of the License, or (at your option) any
   later version.

   zdkimfilter is distributed in the hope that it will be useful, but
   WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the [31]GNU
   General Public Licence for more details.

   As an additional permission under GNU GPLv3 section 7,

     If you modify zdkimfilter, or any covered part of it, by linking or
     combining it with OpenSSL, OpenDKIM, Sendmail, or any software
     developed by The Trusted Domain Project or Sendmail Inc., containing
     parts covered by the applicable licence, the licensor of zdkimfilter
     grants you additional permission to convey the resulting work.

   Copyright (C) 2012-2014 Alessandro Vesely

References

   1. http://www.tana.it/sw/zdkimfilter/v-0.5.shtml
   2. http://www.tana.it/sw/zdkimfilter/zdkimfilter.html
   3. http://www.tana.it/sw/zdkimfilter/dkimsign.html
   4. http://www.tana.it/sw/zdkimfilter/redact.html
   5. http://www.tana.it/sw/zdkimfilter/zaggregate.html
   6. http://www.tana.it/sw/zdkimfilter/zfilter_db.html
   7. http://www.tana.it/sw/zdkimfilter/zdkimfilter.conf.html
   8. http://www.tana.it/sw/zdkimfilter/release-notes-1.5.txt
   9. http://www.tana.it/sw/zdkimfilter/zdkimfilter-1.4.tar.gz
  10. http://www.tana.it/sw/zdkimfilter/zdkimfilter-1.4.tar.gz.asc
  11. http://www.mail-archive.com/courier-users@lists.sourceforge.net/msg37538.html
  12. http://www.tana.it/sw/zdkimfilter/release-notes-1.4.txt
  13. http://www.tana.it/sw/zdkimfilter/zdkimfilter-1.3.tar.gz
  14. http://www.tana.it/sw/zdkimfilter/zdkimfilter-1.3.tar.gz.asc
  15. http://www.mail-archive.com/courier-users@lists.sourceforge.net/msg37523.html
  16. http://www.tana.it/sw/zdkimfilter/release-notes-1.3.txt
  17. http://www.tana.it/sw/zdkimfilter/zdkimfilter-1.2.tar.gz
  18. http://www.tana.it/sw/zdkimfilter/zdkimfilter-1.2.tar.gz.asc
  19. http://www.mail-archive.com/courier-users@lists.sourceforge.net/msg36596.html
  20. http://www.tana.it/sw/zdkimfilter/release-notes-1.2.txt
  21. http://www.tana.it/sw/zdkimfilter/zdkimfilter-1.1.tar.gz
  22. http://www.tana.it/sw/zdkimfilter/zdkimfilter-1.1.tar.gz.asc
  23. http://www.mail-archive.com/courier-users@lists.sourceforge.net/msg36438.html
  24. http://www.tana.it/sw/zdkimfilter/release-notes-1.1.txt
  25. http://www.tana.it/sw/zdkimfilter/zdkimfilter-1.0.tar.gz
  26. http://www.tana.it/sw/zdkimfilter/zdkimfilter-1.0.tar.gz.asc
  27. http://www.mail-archive.com/courier-users@lists.sourceforge.net/msg36402.html
  28. http://www.tana.it/sw/zdkimfilter/v-0.5.shtml
  29. http://www.gnu.org/licenses/gpl.html
  30. http://www.gnu.org/licenses/gpl.html
  31. http://www.gnu.org/licenses/gpl.html


ChangeLog
------------------------------------------------------------------------
r136 | ale | 2015-04-20 13:10:30 +0200 (Mon, 20 Apr 2015) | 8 lines

myadsp: more reliable error codes for TXT lookup failures,
        clearer distinction between DMARC/ADSP policies.
database: NULL domain flags yield zeroes.
zdkimfilter: log when DMARC or ADSP are enabled (verbose>=8),
             log DMARC/ADSP lookup (verbose>=7)
             pretend ADSP policy on NXDOMAIN,
             log INTERNAL ERROR.

------------------------------------------------------------------------
r135 | ale | 2015-04-18 16:09:05 +0200 (Sat, 18 Apr 2015) | 2 lines

tag v1.5

------------------------------------------------------------------------
r134 | ale | 2015-04-18 16:08:43 +0200 (Sat, 18 Apr 2015) | 2 lines

old tag was not exact

------------------------------------------------------------------------
r133 | ale | 2015-04-18 15:56:03 +0200 (Sat, 18 Apr 2015) | 1 line


------------------------------------------------------------------------
r132 | ale | 2015-04-18 15:43:28 +0200 (Sat, 18 Apr 2015) | 7 lines

database: move get_dkim_result() from .c to .h,
          define empty db_set_org_domain() if no OpenDBX.
zdkimfilter: #define dkim_policy around dkim.h (also in myreputation),
             #if-out duplicated code in print_signature_resinfo(),
             return dkim_none for non-processed signatures,
             verify all signing domains if trusted vouchers.

------------------------------------------------------------------------
r131 | ale | 2015-04-17 20:05:56 +0200 (Fri, 17 Apr 2015) | 2 lines

testsuite: comment out DNS real query.

------------------------------------------------------------------------
r130 | ale | 2015-04-17 19:27:07 +0200 (Fri, 17 Apr 2015) | 2 lines

tag v1.5

------------------------------------------------------------------------
r129 | ale | 2015-04-17 19:26:52 +0200 (Fri, 17 Apr 2015) | 3 lines

zdkimfilter: make do_{dmarc,adsp} int, not uint8_t.
docs: minor changes to zdkimfilter(8)

------------------------------------------------------------------------
r128 | ale | 2015-04-17 19:25:01 +0200 (Fri, 17 Apr 2015) | 2 lines

tag v1.5

------------------------------------------------------------------------
r127 | ale | 2015-04-17 16:14:29 +0200 (Fri, 17 Apr 2015) | 5 lines

zdkimfilter: have dmarc_dkim=none if no signatures, ditto spf.
docs: complete zaggregate man page.
examples: remove unneeded escapes.


------------------------------------------------------------------------
r126 | ale | 2015-04-15 18:17:00 +0200 (Wed, 15 Apr 2015) | 10 lines

database: fix prefix_len and adsp.
zdkimfilter: move BADSIG out of logged permerrors,
             review use of dmarc reason and disposition,
             log (verbose>=3) if BOFHSPFFROM is the sole pass.
testsuite: account for the aded log line.
configure: add --enable-dkimsign-setuid.
docs: finish query documentation in zfilter_db,
      mention unauthorized scope in zdkimfilter.conf,
      start zaggregate man page.

------------------------------------------------------------------------
r125 | ale | 2015-04-11 09:43:53 +0200 (Sat, 11 Apr 2015) | 3 lines

zdkimfilter: set dkim_order according to preference.
database: compute prefix_len for non-aligned domains.

------------------------------------------------------------------------
r124 | ale | 2015-04-11 08:54:51 +0200 (Sat, 11 Apr 2015) | 6 lines

zdkimfilter: new functions copy_replacement(),
             replace_received_auth(), and
             replace_courier_wrap().
rfc822: imported Courier sources to parse and rewrite addresses.
testsuite: check wrapping long lines on signing.

------------------------------------------------------------------------
r123 | ale | 2015-04-10 10:18:31 +0200 (Fri, 10 Apr 2015) | 5 lines

zaggregate: fix limit comparison.
zdkimfilter: log_dkim_order_above as warning.
example: review, fix table creation.
docs: added new z parms, begin zfilter_db.

------------------------------------------------------------------------
r122 | ale | 2015-04-07 13:01:22 +0200 (Tue, 07 Apr 2015) | 10 lines

zdkimfilter: clarify use of dkim_permerror vs dkim_fail,
             better verification failure messages,
             define default report interval in database.h,
             change default from 3600 to 86400.
database: add original_ri besides adjusted dmarc_ri.
zaggregate: fix xml structure, allow adding reason comment,
            choose what spf scope to report.
example: add 'none' defaults, clean up odbx_example.sql and
         odbx_example.sh, add zaggregate-example.sh.

------------------------------------------------------------------------
r121 | ale | 2015-04-02 18:44:28 +0200 (Thu, 02 Apr 2015) | 29 lines

zaggregate: new program to produce DMARC reports.
cstring: new source to ease handling C strings.
database: new statement db_sql_set_dmarc_agg,
          new vars period, period_start, and period_end,
          flag variable bumped to 64 bit,
          #define CONSOLE_DEBUG to display queries,
          new function db_zag_wrapup() for dmarc only,
          modified stmt_run_n() for multiple rows,
          added 3 public functions for dmarc queries,
          for zfilter_db, replace comma with slash,
          run adjust_rua() against test records.
zdkimfilter: new params honored_report_interval and log_dkim_order_above,
             check report interval, aka period, is 24h submultiple,
             log if number of signing domains exceeds order,
             adjust rua and ri before passing them to the database.
filterlib: define live_children volatile.
myadsp: fix dmarc record parsing,
        new function adjust_ri() to check submultiples,
        new function adjust_rua() to parse dmarc URI,
        have write_dmarc_rec() use the new functions,
        new function verify_dmarc_addr() for external recipients.
odbx_example: add DMARC queries,
              add odbx_example.sh with calls to zfilter.db.
configure: back to version 1.5,

src/md5.c
/* MD5C.C - RSA Data Security, Inc., MD5 message-digest algorithm
 */

/* Copyright (C) 1991-2, RSA Data Security, Inc. Created 1991. All
rights reserved.

License to copy and use this software is granted provided that it
is identified as the "RSA Data Security, Inc. MD5 Message-Digest
Algorithm" in all material mentioning or referencing this software
or this function.

License is also granted to make and use derivative works provided
that such works are identified as "derived from the RSA Data
Security, Inc. MD5 Message-Digest Algorithm" in all material
mentioning or referencing the derived work.

RSA Data Security, Inc. makes no representations concerning either
the merchantability of this software or the suitability of this
software for any particular purpose. It is provided "as is"
without express or implied warranty of any kind.

These notices must be retained in any copies of any part of this
documentation and/or software.
 */

#include "md5.h"

/* Constants for MD5Transform routine.
 */

#define S11 7
#define S12 12
#define S13 17
#define S14 22
#define S21 5
#define S22 9
#define S23 14
#define S24 20
#define S31 4
#define S32 11
#define S33 16
#define S34 23
#define S41 6
#define S42 10
#define S43 15
#define S44 21

/* POINTER defines a generic pointer type */
typedef unsigned char *POINTER;

static void MD5Transform(uint32_t [4], unsigned char [64]);
static void Encode(unsigned char *, uint32_t *, unsigned int);
static void Decode(uint32_t *, unsigned char *, unsigned int);
#if 0
static void MD5_memcpy(POINTER, POINTER, unsigned int);
static void MD5_memset(POINTER, int, unsigned int);
#endif
#include <string.h>
#define MD5_memcpy memcpy
#define MD5_memset memset

static unsigned char PADDING[64] = {
  0x80, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
  0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
  0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0
};

/* F, G, H and I are basic MD5 functions.
 */
#define F(x, y, z) (((x) & (y)) | ((~x) & (z)))
#define G(x, y, z) (((x) & (z)) | ((y) & (~z)))
#define H(x, y, z) ((x) ^ (y) ^ (z))
#define I(x, y, z) ((y) ^ ((x) | (~z)))

/* ROTATE_LEFT rotates x left n bits.
 */
#define ROTATE_LEFT(x, n) (((x) << (n)) | ((x) >> (32-(n))))

/* FF, GG, HH, and II transformations for rounds 1, 2, 3, and 4.
Rotation is separate from addition to prevent recomputation.
 */
#define FF(a, b, c, d, x, s, ac) { \
 (a) += F ((b), (c), (d)) + (x) + (uint32_t)(ac); \
 (a) = ROTATE_LEFT ((a), (s)); \
 (a) += (b); \
  }
#define GG(a, b, c, d, x, s, ac) { \
 (a) += G ((b), (c), (d)) + (x) + (uint32_t)(ac); \
 (a) = ROTATE_LEFT ((a), (s)); \
 (a) += (b); \
  }
#define HH(a, b, c, d, x, s, ac) { \
 (a) += H ((b), (c), (d)) + (x) + (uint32_t)(ac); \
 (a) = ROTATE_LEFT ((a), (s)); \
 (a) += (b); \
  }
#define II(a, b, c, d, x, s, ac) { \
 (a) += I ((b), (c), (d)) + (x) + (uint32_t)(ac); \
 (a) = ROTATE_LEFT ((a), (s)); \
 (a) += (b); \
  }

/* MD5 initialization. Begins an MD5 operation, writing a new context.
 */
void MD5Init (MD5_CTX *context)
{
  context->count[0] = context->count[1] = 0;
  /* Load magic initialization constants.
*/
  context->state[0] = 0x67452301;
  context->state[1] = 0xefcdab89;
  context->state[2] = 0x98badcfe;
  context->state[3] = 0x10325476;
}

/* MD5 block update operation. Continues an MD5 message-digest
  operation, processing another message block, and updating the
  context.
 */
void MD5Update (MD5_CTX *context, unsigned char *input, unsigned int inputLen)
{
  unsigned int i, index, partLen;

  /* Compute number of bytes mod 64 */
  index = (unsigned int)((context->count[0] >> 3) & 0x3F);

  /* Update number of bits */
  if ((context->count[0] += ((uint32_t)inputLen << 3))
   < ((uint32_t)inputLen << 3))
     context->count[1]++;
  context->count[1] += ((uint32_t)inputLen >> 29);

  partLen = 64 - index;

  /* Transform as many times as possible.
*/
  if (inputLen >= partLen) {
 MD5_memcpy
   ((POINTER)&context->buffer[index], (POINTER)input, partLen);
 MD5Transform (context->state, context->buffer);

 for (i = partLen; i + 63 < inputLen; i += 64)
   MD5Transform (context->state, &input[i]);

 index = 0;
  }
  else
 i = 0;

  /* Buffer remaining input */
  MD5_memcpy
 ((POINTER)&context->buffer[index], (POINTER)&input[i],
  inputLen-i);
}

/* MD5 finalization. Ends an MD5 message-digest operation, writing the
  the message digest and zeroizing the context.
 */
void MD5Final (unsigned char digest[16], MD5_CTX *context)
{
  unsigned char bits[8];
  unsigned int index, padLen;

  /* Save number of bits */
  Encode (bits, context->count, 8);

  /* Pad out to 56 mod 64.
*/
  index = (unsigned int)((context->count[0] >> 3) & 0x3f);
  padLen = (index < 56) ? (56 - index) : (120 - index);
  MD5Update (context, PADDING, padLen);

  /* Append length (before padding) */
  MD5Update (context, bits, 8);

  /* Store state in digest */
  Encode (digest, context->state, 16);

  /* Zeroize sensitive information.
*/
  MD5_memset ((POINTER)context, 0, sizeof (*context));
}

/* MD5 basic transformation. Transforms state based on block.
 */
static void MD5Transform (uint32_t state[4], unsigned char block[64])
{
  uint32_t a = state[0], b = state[1], c = state[2], d = state[3], x[16];

  Decode (x, block, 64);

  /* Round 1 */
  FF (a, b, c, d, x[ 0], S11, 0xd76aa478); /* 1 */
  FF (d, a, b, c, x[ 1], S12, 0xe8c7b756); /* 2 */
  FF (c, d, a, b, x[ 2], S13, 0x242070db); /* 3 */
  FF (b, c, d, a, x[ 3], S14, 0xc1bdceee); /* 4 */
  FF (a, b, c, d, x[ 4], S11, 0xf57c0faf); /* 5 */
  FF (d, a, b, c, x[ 5], S12, 0x4787c62a); /* 6 */
  FF (c, d, a, b, x[ 6], S13, 0xa8304613); /* 7 */
  FF (b, c, d, a, x[ 7], S14, 0xfd469501); /* 8 */
  FF (a, b, c, d, x[ 8], S11, 0x698098d8); /* 9 */
  FF (d, a, b, c, x[ 9], S12, 0x8b44f7af); /* 10 */
  FF (c, d, a, b, x[10], S13, 0xffff5bb1); /* 11 */
  FF (b, c, d, a, x[11], S14, 0x895cd7be); /* 12 */
  FF (a, b, c, d, x[12], S11, 0x6b901122); /* 13 */
  FF (d, a, b, c, x[13], S12, 0xfd987193); /* 14 */
  FF (c, d, a, b, x[14], S13, 0xa679438e); /* 15 */
  FF (b, c, d, a, x[15], S14, 0x49b40821); /* 16 */

 /* Round 2 */
  GG (a, b, c, d, x[ 1], S21, 0xf61e2562); /* 17 */
  GG (d, a, b, c, x[ 6], S22, 0xc040b340); /* 18 */
  GG (c, d, a, b, x[11], S23, 0x265e5a51); /* 19 */
  GG (b, c, d, a, x[ 0], S24, 0xe9b6c7aa); /* 20 */
  GG (a, b, c, d, x[ 5], S21, 0xd62f105d); /* 21 */
  GG (d, a, b, c, x[10], S22,  0x2441453); /* 22 */
  GG (c, d, a, b, x[15], S23, 0xd8a1e681); /* 23 */
  GG (b, c, d, a, x[ 4], S24, 0xe7d3fbc8); /* 24 */
  GG (a, b, c, d, x[ 9], S21, 0x21e1cde6); /* 25 */
  GG (d, a, b, c, x[14], S22, 0xc33707d6); /* 26 */
  GG (c, d, a, b, x[ 3], S23, 0xf4d50d87); /* 27 */
  GG (b, c, d, a, x[ 8], S24, 0x455a14ed); /* 28 */
  GG (a, b, c, d, x[13], S21, 0xa9e3e905); /* 29 */
  GG (d, a, b, c, x[ 2], S22, 0xfcefa3f8); /* 30 */
  GG (c, d, a, b, x[ 7], S23, 0x676f02d9); /* 31 */
  GG (b, c, d, a, x[12], S24, 0x8d2a4c8a); /* 32 */

  /* Round 3 */
  HH (a, b, c, d, x[ 5], S31, 0xfffa3942); /* 33 */
  HH (d, a, b, c, x[ 8], S32, 0x8771f681); /* 34 */
  HH (c, d, a, b, x[11], S33, 0x6d9d6122); /* 35 */
  HH (b, c, d, a, x[14], S34, 0xfde5380c); /* 36 */
  HH (a, b, c, d, x[ 1], S31, 0xa4beea44); /* 37 */
  HH (d, a, b, c, x[ 4], S32, 0x4bdecfa9); /* 38 */
  HH (c, d, a, b, x[ 7], S33, 0xf6bb4b60); /* 39 */
  HH (b, c, d, a, x[10], S34, 0xbebfbc70); /* 40 */
  HH (a, b, c, d, x[13], S31, 0x289b7ec6); /* 41 */
  HH (d, a, b, c, x[ 0], S32, 0xeaa127fa); /* 42 */
  HH (c, d, a, b, x[ 3], S33, 0xd4ef3085); /* 43 */
  HH (b, c, d, a, x[ 6], S34,  0x4881d05); /* 44 */
  HH (a, b, c, d, x[ 9], S31, 0xd9d4d039); /* 45 */
  HH (d, a, b, c, x[12], S32, 0xe6db99e5); /* 46 */
  HH (c, d, a, b, x[15], S33, 0x1fa27cf8); /* 47 */
  HH (b, c, d, a, x[ 2], S34, 0xc4ac5665); /* 48 */

  /* Round 4 */
  II (a, b, c, d, x[ 0], S41, 0xf4292244); /* 49 */
  II (d, a, b, c, x[ 7], S42, 0x432aff97); /* 50 */
  II (c, d, a, b, x[14], S43, 0xab9423a7); /* 51 */
  II (b, c, d, a, x[ 5], S44, 0xfc93a039); /* 52 */
  II (a, b, c, d, x[12], S41, 0x655b59c3); /* 53 */
  II (d, a, b, c, x[ 3], S42, 0x8f0ccc92); /* 54 */
  II (c, d, a, b, x[10], S43, 0xffeff47d); /* 55 */
  II (b, c, d, a, x[ 1], S44, 0x85845dd1); /* 56 */
  II (a, b, c, d, x[ 8], S41, 0x6fa87e4f); /* 57 */
  II (d, a, b, c, x[15], S42, 0xfe2ce6e0); /* 58 */
  II (c, d, a, b, x[ 6], S43, 0xa3014314); /* 59 */
  II (b, c, d, a, x[13], S44, 0x4e0811a1); /* 60 */
  II (a, b, c, d, x[ 4], S41, 0xf7537e82); /* 61 */
  II (d, a, b, c, x[11], S42, 0xbd3af235); /* 62 */
  II (c, d, a, b, x[ 2], S43, 0x2ad7d2bb); /* 63 */
  II (b, c, d, a, x[ 9], S44, 0xeb86d391); /* 64 */

  state[0] += a;
  state[1] += b;
  state[2] += c;
  state[3] += d;

  /* Zeroize sensitive information.*/
  MD5_memset ((POINTER)x, 0, sizeof (x));
}

/* Encodes input (uint32_t) into output (unsigned char). Assumes len is
  a multiple of 4.
 */
static void Encode (unsigned char *output, uint32_t *input, unsigned int len)
{
  unsigned int i, j;

  for (i = 0, j = 0; j < len; i++, j += 4) {
    output[j] = (unsigned char)(input[i] & 0xff);
    output[j+1] = (unsigned char)((input[i] >> 8) & 0xff);
    output[j+2] = (unsigned char)((input[i] >> 16) & 0xff);
    output[j+3] = (unsigned char)((input[i] >> 24) & 0xff);
  }
}

/* Decodes input (unsigned char) into output (uint32_t). Assumes len is
  a multiple of 4.
 */
static void Decode (uint32_t *output, unsigned char *input, unsigned int len)
{
  unsigned int i, j;

  for (i = 0, j = 0; j < len; i++, j += 4)
    output[i] = ((uint32_t)input[j]) | (((uint32_t)input[j+1]) << 8) |
      (((uint32_t)input[j+2]) << 16) | (((uint32_t)input[j+3]) << 24);
}

#if 0
/* Note: Replace "for loop" with standard memcpy if possible.
 */

static void MD5_memcpy (output, input, len)
POINTER output;
POINTER input;
unsigned int len;
{
  unsigned int i;

  for (i = 0; i < len; i++)
    output[i] = input[i];
}

/* Note: Replace "for loop" with standard memset if possible.
 */
static void MD5_memset (output, value, len)
POINTER output;
int value;
unsigned int len;
{
  unsigned int i;

  for (i = 0; i < len; i++)
    ((char *)output)[i] = (char)value;
}
#endif
